instrumentation/grpc-1.5/library/src/main/java/io/grpc/override/ContextStorageOverride.java

```diff
@@ -0,0 +1,122 @@
+import java.util.Deque;
+import java.util.Map;
+import java.util.WeakHashMap;
+import java.util.logging.Level;
+import java.util.logging.Logger;
+
+/**
+ * {@link Context.Storage} override which uses OpenTelemetry context as the backing store. Both gRPC
+ * and OpenTelemetry contexts refer to each other to ensure that both OTel context propagation
+ * mechanisms and gRPC context propagation mechanisms can be used interchangably.
+ */
+public final class ContextStorageOverride extends Context.Storage {
+
+  private static final Logger logger = Logger.getLogger(ContextStorageOverride.class.getName());
+
+  private static final ContextKey<Context> GRPC_CONTEXT = ContextKey.named("grpc-context");
+  private static final Context.Key<io.opentelemetry.context.Context> OTEL_CONTEXT =
+      Context.key("otel-context");
+
+  // Because the extension point is void, there is no way to return information about the backing
+  // OpenTelemetry context when attaching gRPC context. So the only option is to have this
+  // side-channel to keep track of scopes. Because the same context can be attached to multiple
+  // threads, we must use a ThreadLocal here - on the bright side it means the map doesn't have to
+  // be concurrent. This will add an additional threadlocal lookup when attaching / detaching gRPC
+  // context, but not when accessing the current. In many applications, this means a small
+  // difference
+  // since those operations are rare, but in highly reactive applications where the overhead of
+  // ThreadLocal was already a problem, this makes it worse.
+  private static final ThreadLocal<WeakHashMap<Context, Deque<Scope>>> contextScopes =
+      ThreadLocal.withInitial(WeakHashMap::new);
+
+  @Override
+  public void attach(Context toAttach) {
+    io.opentelemetry.context.Context otelContext = io.opentelemetry.context.Context.current();
+    Context current = otelContext.get(GRPC_CONTEXT);
+
+    if (current == toAttach) {
+      contextScopes
+          .get()
+          .computeIfAbsent(toAttach, unused -> new ArrayDeque<>())
+          .addLast(Scope.noop());
+      return;
+    }
+
+    io.opentelemetry.context.Context base = OTEL_CONTEXT.get(toAttach);
+    final io.opentelemetry.context.Context newOtelContext;
+    if (base != null) {
+      // gRPC context which has an OTel context associated with it via a call to
+      // ContextStorageOverride.current(). Using it as the base allows it to be propagated together
+      // with the gRPC context.
+      newOtelContext = base.with(GRPC_CONTEXT, toAttach);
+    } else {
+      // gRPC context without an OTel context associated with it. This is only possible when
+      // attaching a context directly created by Context.ROOT, e.g., Context.ROOT.with(...) which
+      // is not common. We go ahead and assume the OTel context should also be reset to root in this
+      // case.
+      newOtelContext = io.opentelemetry.context.Context.root().with(GRPC_CONTEXT, toAttach);
```

Comment:
what do you think of erroring on the side of not breaking otel context in this case and using otel `Context.current()` here? e.g. I'm guessing one use case for grpc Context.ROOT.attach() is at request entry, as a safety for not leaking context, similar to what we do with otel context

Reply:
Yeah - that's why I thought we should reset the otel context too, it seemed like you wouldn't ever reset one and not the other, similar to how we always propagate them together.

Reply:
but in that case, it will break our otel propagation, b/c the agent attaches the SERVER span to the otel context even earlier in the request entry

Reply:
I guess that makes sense - and ideally we just don't leak context anyways so it doesn't matter too much :D